modules/lang-painless/src/main/java/org/elasticsearch/painless/api/Augmentation.java

```diff
@@ -552,4 +554,138 @@ public static String decodeBase64(String receiver) {
         // O(N) or faster depending on implementation
         return result.toArray(new String[0]);
     }
+
+
+    // getByPath public API
+    /**
+     * Access values in nested containers with a dot separated path.  Path elements are treated
+     * as strings for Maps and integers for Lists.
+     * Throws 'IllegalArgumentException' if path does not exist or a path element for a List is not an integer.
+     */
+    public static <E> Object getByPath(List<E> receiver, String path) {
+        return getByPath(receiver, splitPath(path), 0, getByPathMissing(path));
+    }
+
+    /**
+     * Access values in nested containers with a dot separated path.  Path elements are treated
+     * as strings for Maps and integers for Lists.
+     * Throws 'IllegalArgumentException' if path does not exist or a path element for a List is not an integer.
+     */
+    public static <K,V> Object getByPath(Map<K,V> receiver, String path) {
+        return getByPath(receiver, splitPath(path), 0, getByPathMissing(path));
+    }
+
+    /**
+     * Same as 'getByPath(List, String)' but returns defaultValue if path does not exist.
+     * Throws 'IllegalArgumentException' if there is a non-container type at a non-terminal path element.
+     */
+    public static <E> Object getByPath(List<E> receiver, String path, Object defaultValue) {
+        return getByPath(receiver, splitPath(path), 0, () -> defaultValue);
+    }
+
+    /**
+     * Same as 'getByPath(Map, String)' but returns defaultValue if path does not exist.
+     * Throws 'IllegalArgumentException' if there is a non-container type at a non-terminal path element.
+     */
+    public static <K,V> Object getByPath(Map<K,V> receiver, String path, Object defaultValue) {
+        return getByPath(receiver, splitPath(path), 0, () -> defaultValue);
+    }
+
+    // getByPath helpers
+    /**
+     * Dispatch function for getByPath, handling successful lookup, obj as Map, List or non-collection.
+     * Throws 'IllegalArgumentException' if elements[i] is an empty string, indicating double separators
+     * or path ends in a separator.
+     */
+    private static Object getByPath(Object obj, String[] elements, int i, Supplier<Object> defaultSupplier) {
```

Comment:
I think it would be more clear with each of these names having one more word. Eg Dispatch here, Map/List below.